src/Orleans.Core/Messaging/MessageSerializer.cs

```diff
@@ -317,10 +317,18 @@ internal List<GrainAddressCacheUpdate> ReadCacheInvalidationHeaders<TInput>(ref
             return new List<GrainAddressCacheUpdate>();
         }
 
-        internal void WriteCacheInvalidationHeaders<TBufferWriter>(ref Writer<TBufferWriter> writer, List<GrainAddressCacheUpdate> value) where TBufferWriter : IBufferWriter<byte>
+        internal void WriteCacheInvalidationHeaders<TBufferWriter>(ref Writer<TBufferWriter> writer, Message message) where TBufferWriter : IBufferWriter<byte>
         {
-            writer.WriteVarUInt32((uint)value.Count);
-            foreach (var entry in value)
+            // Create a snapshot to avoid concurrent modification exceptions during enumeration.
+            // The list can be modified by other threads after the message is queued for sending.
+            GrainAddressCacheUpdate[] snapshot;
+            lock (message.CacheInvalidationHeaderLock)
```

Comment:
Put the lock around the whole body, rather than locking just this. Don't use ToArray, just enumerate directly under the lock